### PR TITLE
Feature/fix rlp push0

### PR DESCRIPTION
**Pull request by laisolizq**
- Add push0
- Fix RLP data length

### CODE REVIEW

File: main/utils.zkasm [hunk 1]
```diff
@@ -872,10 +872,13 @@ SHLarithfinal:
 
 ; out of counters full tracer event trigger
 outOfCountersStep:
+    $                           :MLOAD(isLoadRLP),JMPNZ(handleOOCSatRLP)
```
Comment (ignasirv) [inline main/utils.zkasm:875]: What do you think adding this check only once at `handleBatchError` function?
**$                           :MLOAD(isLoadRLP),JMPNZ(handleOOCatRLP)**
And remove the `handleOOCXatRLP` from the code

File: main/vars.zkasm [hunk 2]
```diff
@@ -90,4 +90,5 @@ VAR CTX salt ; CREATE2 parameter 'salt' used to compute new contract address
 VAR CTX gasCTX ; remaining gas in the origin CTX when a new context is created
 VAR CTX dataStarts; hash position where de transaction 'data' starts in the batchHashData
 VAR CTX isPreEIP155 ; flag to check if the current tx is legacy, previous to Spurious Dragon (EIP-155)
-VAR CTX initTouchedSR ; touched root once a new context begins
\ No newline at end of file
+VAR CTX initTouchedSR ; touched root once a new context begins
+VAR CTX isLoadRLP ; flag to determine if the function is called from RLP loop
```
Comment (ignasirv) [inline main/vars.zkasm:94]: Change to isLoadingRLP or isProcessingRLP

File: main/main.zkasm [hunk 3]
```diff
@@ -106,11 +106,15 @@ skipSetGlobalExitRoot:
...
         A+1 => CTX      :MSTORE(lastCtxUsed)
+        ; set flag isLoadRLP to 1
+        1               :MSTORE(isLoadRLP)
 
         $ => A          :MLOAD(batchL2DataLength)
         $ => C          :MLOAD(batchL2DataParsed)
         C - A           :JMPN(loadTx_rlp, endCheckRLP)
 endCheckRLP:
+        ; set flag isLoadRLP to 0
+        0               :MSTORE(isLoadRLP)
```
Comment (ignasirv) [inline main/main.zkasm:117]: The var now is of type CTX. We have to options:
- Set the flag to 0 before calling `loadTx_rlp`: this way, all the CTX are set to 0
- Make the var GLOBAL: in this case we should set it to 1 only once just before `txLoopRLP`
The current situation is that we are only setting the flag to 0 for the last processed context

File: main/utils.zkasm [hunk 4]
```diff
@@ -1153,6 +1156,7 @@ finalPush:
 VAR GLOBAL tmpVarDaddB
 VAR GLOBAL tmpZkPCaddB
 VAR GLOBAL auxBytes
+VAR GLOBAL auxLengthBytes
```
[the remark this answers is not in the document]
Comment (krlosMata) [reply main/utils.zkasm:1159]: I think we can compute in `invalidTxRLP`:
- `batchHashPos` - `batchL2DataLength` to check how many bytes are missing to fill in the the `batchHashData`
- so we can save the `auxLengthBytes` variable

**Review by krlosMata — CHANGES_REQUESTED** [review]

**Review by abdulmenan10 — APPROVED** [review]

File: main/utils.zkasm (hunk 4 above)
[the remark this answers is not in the document]
Comment (abdulmenan10) [reply main/utils.zkasm]: Yes

**Comment by abdulmenan10** [comment]
Yes

**Review by ignasirv — APPROVED** [review]
💯

**Review by invocamanman — APPROVED** [review]

**Review by krlosMata — APPROVED** [review]
:smile_cat:

File: main/opcodes/stack-operations.zkasm [hunk 5]
```diff
@@ -7,6 +7,15 @@
  *  - stack input: none
  *  - stack output: [pushed_value]
  */
+
+opPUSH0:
```
Comment (gumb0) [inline main/opcodes/stack-operations.zkasm:11]: Does it not consume any ZK counters? I would think that gas check at stack check do. Even POP requires 100 steps.

File: main/opcodes/stack-operations.zkasm [hunk 6]
```diff
@@ -7,6 +7,15 @@
  *  - stack input: none
  *  - stack output: [pushed_value]
  */
```
Comment (gumb0) [inline main/opcodes/stack-operations.zkasm:9]: This comment applies to `opPUSH1` and further, `opPUSH0` should be above it and with its own comment.

File: main/load-tx-rlp.zkasm [hunk 7]
```diff
@@ -290,30 +290,22 @@ vREADTx:
...
-handleOOCSatRLP:
-        $${eventLog(onError, OOCS)}      :JMP(handleOOCatRLP)
-handleOOCatRLP:
-        $ => SR                          :MLOAD(batchSR)
-
 invalidTxRLP:
 ;; Append all missing 'batchL2Data' to 'batchDataHash' bytes
         $ => B                          :MLOAD(batchL2DataLength)
         $ => C                          :MLOAD(batchHashPos)
+        $${p = C}
```
Comment (gumb0) [inline main/load-tx-rlp.zkasm:308]: Why is this needed? Seems like it can change behaviour (of `getTxs()`?) significantly

File: main/utils.zkasm [hunk 8]
```diff
@@ -928,6 +928,7 @@ handleError:
 handleBatchError:
     ; restore init state root and finish batch
     $ => SR         :MLOAD(batchSR)
+    $               :MLOAD(isLoadingRLP),JMPNZ(invalidTxRLP)
```
Comment (gumb0) [inline main/utils.zkasm:931]: Would be good to have a comment why RLP error processing is different

File: main/load-tx-rlp.zkasm (hunk 7 above)
Comment (krlosMata) [reply main/load-tx-rlp.zkasm:308]: `p` is the pointer to the read transactions bytes via `getTxs()`.

While reading RLP, we add bytes to the `batchHash` and perform some check afterwards, but the bytes are already added. So, if anything goes wrong we jump to `invalidTxRLP` to load the remaining bytes. The remaining bytes where added starting at `p`, where the `p` was set during the RLP parsing. This was always true before, but if we perform the check before the bytes are added to the `batchHash`, then it will not match the `batchHash`.ç

The generic way to add the missing `batchHash` bytes was to simply set the pointer `p` to starting byte to add, which is stored in `batchHashPos`

**Comment by gumb0** [comment]
Some general thoughts about RLP fix, maybe missing the entire picture and all implications.

In general it feels like fixing the problem of managing one global thing (ZK counters, behaving differently in different contexts), with adding more global data. Overall this increases complexity and potential implicit interdependencies and difficulty of reasoning about code.

Ideally I think it would be better to make this dependency explicit by e.g. having an extra flag parameter to those `utils` that are called from RLP processing and should process counter errors differently.

I can see that this might be impractical if those `utils` are also called from many other places. Another less radical suggestion to consider could be: do the branching based on global `isLoadingRLP` only inside requried `utils` instead of lower-level `handleBatchError`. This way the fix would affect only some limited scope of functionality, comparing to affecting every batch error processing like now. And it would still be more explicit which utils depend on this global state.
(Downside of this approach could be if RLP processing changes to using some new `utils`, you have to remember to check `isLoadingRLP` in each one)

File: main/opcodes/stack-operations.zkasm (hunk 5 above)
Comment (krlosMata) [reply main/opcodes/stack-operations.zkasm:11]: Yes, we would add them

File: main/opcodes/stack-operations.zkasm (hunk 6 above)
Comment (krlosMata) [reply main/opcodes/stack-operations.zkasm:9]: Agree !! We will add specific comment for `PUSH0`

File: main/load-tx-rlp.zkasm (hunk 7 above)
Comment (krlosMata) [reply main/load-tx-rlp.zkasm:308]: > So as I understand this error processing is now called from more places (`utils` generating out-of-counters errors), including from the points where `p` is not updated yet, so it is initialized explicitly.

exactly!

Comment (krlosMata) [reply main/load-tx-rlp.zkasm:308]: > > The remaining bytes where added starting at `p`, where the `p` was set during the RLP parsing.
> 
> But for old cases, where this error processing is called after `p` is already set, wouldn't this code here add more bytes than needed?

for old cases, new code will just only set the `p` where it was already read, so it will just only rewrite the `p` with same value

**Comment by krlosMata** [comment]
> Some general thoughts about RLP fix, maybe missing the entire picture and all implications.
> 
> In general it feels like fixing the problem of managing one global thing (ZK counters, behaving differently in different contexts), with adding more global data. Overall this increases complexity and potential implicit interdependencies and difficulty of reasoning about code.
> 
> Ideally I think it would be better to make this dependency explicit by e.g. having an extra flag parameter to those `utils` that are called from RLP processing and should process counter errors differently.
> 
> I can see that this might be impractical if those `utils` are also called from many other places. Another less radical suggestion to consider could be: do the branching based on global `isLoadingRLP` only inside requried `utils` instead of lower-level `handleBatchError`. This way the fix would affect only some limited scope of functionality, comparing to affecting every batch error processing like now. And it would still be more explicit which utils depend on this global state. (Downside of this approach could be if RLP processing changes to using some new `utils`, you have to remember to check `isLoadingRLP` in each one)

We decided to go with the `isLoadingRLP` approach since it povides better security when adding new code to the RLP parsing since it is handled globally and it is not function specific. Also, RLP error where just removed and all zk-counters are handled via `handleBatchError`

File: main/utils.zkasm (hunk 8 above)
Comment (krlosMata) [reply main/utils.zkasm:931]: This flag is not used outside the `zkevm-rom` so it could not cause any issue.
I would say the opposite, it could add information if the limit counters has been reached during the RLP
